server/controller/urlController.js

```diff
@@ -1,12 +1,49 @@
 const urlModel = require("../model/urlSchema");
+const userModel = require("../model/userSchema");
 const validUrl = require("valid-url");
 const uniqueString = require("../utils/utils");
 const dbConnect = require("../model/dbConnect");
 const dotenv = require("dotenv");
+const jwt = require("jsonwebtoken");
 
 dotenv.config();
 
 // Define asynchronous function to handle specific URL retrieval
+const getAllUserUrls = async (req, res) => {
+  try {
+    const { token } = req.params;
+
+    // Connect to the database
+    await dbConnect();
+
+    // Decode the token to extract user._id
+    const decodedToken = jwt.verify(token, process.env.JWT_SECRET);
+    const userId = decodedToken.user._id;
```

Comment:
get userId from req params and id should be in routes

Reply:
what I have done is after authenticating in middleware and extracting data from token , I have put the data in req.user so we can access userId by req.user.userId after authentication.

is this ok? @abhijeetnishal

Reply:
We need middleware so that we can use in other endpoints so that we don't need to add in controller following design principles.
That's why use this approach as mentioned

Reply:
> get userId from req params and id should be in routes

I am not getting this approach

Reply:
I have 

> @atharvarekhawar We need middleware so that we can use in other endpoints so that we don't need to add in controller following design principles. That's why use this approach as mentioned

I have implemented middleware

Reply:
exports.isAuthenticated = async (req, res, next) => {
  try {
    // Extracting JWT from request cookies, body or header
    const token = req.body.token;

    // If JWT is missing, return 401 Unauthorized response
    if (!token) {
      return res.status(401).json({ success: false, message: `Token Missing` });
    }

    try {
      // Verifying the JWT using the secret key stored in environment variables
      const decode = await jwt.verify(token, process.env.JWT_SECRET);
      req.user = decode;
    } catch (error) {
      // If JWT verification fails, return 401 Unauthorized response
      return res
        .status(401)
        .json({ success: false, message: "token is invalid" });
    }

    // If JWT is valid, move on to the next middleware or request handler
    next();
  } catch (error) {
    // If there is an error during the authentication process, return 401 Unauthorized response
    return res.status(401).json({
      success: false,
      message: `Something Went Wrong While Validating the Token`,
    });
  }
};

Reply:
``exports.isAuthenticated = async (req, res, next) => {
  try {
    // Extracting JWT from request cookies, body or header
    const token = req.body.token;

    // If JWT is missing, return 401 Unauthorized response
    if (!token) {
      return res.status(401).json({ success: false, message: `Token Missing` });
    }

    try {
      // Verifying the JWT using the secret key stored in environment variables
      const decode = await jwt.verify(token, process.env.JWT_SECRET);
      req.user = decode;
    } catch (error) {
      // If JWT verification fails, return 401 Unauthorized response
      return res
        .status(401)
        .json({ success: false, message: "token is invalid" });
    }

    // If JWT is valid, move on to the next middleware or request handler
    next();
  } catch (error) {
    // If there is an error during the authentication process, return 401 Unauthorized response
    return res.status(401).json({
      success: false,
      message: `Something Went Wrong While Validating the Token`,
    });
  }
};

Reply:
should I push ? @abhijeetnishal

Reply:
`try {
      // Verifying the JWT using the secret key stored in environment variables
        const decode = await jwt.verify(token, process.env.JWT_SECRET);
        req.userId = decode.id;
    } catch (error) {
      // If JWT verification fails, return 401 Unauthorized response
      return res
        .status(401)
        .json({ success: false, message: "token is invalid" });
    }`

this is how after extracting data in middleware the userId will be passed in req

Reply:
try {
      // Verifying the JWT using the secret key stored in environment variables
        const decode = await jwt.verify(token, process.env.JWT_SECRET);
        req.userId = decode.id;
    } catch (error) {
      // If JWT verification fails, return 401 Unauthorized response
      return res
        .status(401)
        .json({ success: false, message: "token is invalid" });
    }
    
    auserId will be added to req after passing through middleware

Reply:
> > get userId from req params and id should be in routes
> 
> I am not getting this approach

You are getting id from token in controller in you code instead we can use middleware functions directly in express router method so that we don't need to decode this and check for every controller method. 
Basicaaly we are following design principle to reuse the same code.
I hope you understand now.

Reply:
Don't worry in a day or two we will have our discord server channel we will discuss what ever you need help.

Reply:
no, I have changed the controller code

Reply:
const getAllUserUrls = async (req, res) => {
  try {
    // Connect to the database
    await dbConnect();

    const userId = req.userId;

    // Find all URLs created by the user
    const userUrls = await urlModel.find({ userId: userId });

    res.status(200).json({ urls: userUrls });

  } catch (error) {
    // Handle any errors that occur during the process
    console.error(error);
    res.status(500).json({ message: "Internal Server Error" });
  }
};

new controller @abhijeetnishal , should I push?

Reply:
> get userId from req params and id should be in routes

As i mentioned here